### ISSUE TITLE
Home -> ButtonsPaletteChooser: Incorrect Content Reading Order and Labelling 

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#755]
There are some serious problems with the order of content and labelling of controls in this dialog, making it incredibly confusing to use.  Indeed, many users may find it completely unusable. 
 
Each button palette in the dialog has: 
 
* a group title (such as “Operators” or “Notations”) which is not marked up as a heading; 
* a list of buttons, all of which are marked as disabled but still have shortcut keys listed; 
* a check box, the purpose of which is unknown because it has no accessible label; and 
* a heading with the group title. 
 
The above list reflects the order in which a screen reader user will encounter the content and controls.  Note that the heading is positioned after the content to which it relates.  Due to the presence of multiple button rows, a screen reader user is almost guaranteed to either miss content when navigating by heading, or become confused as to which buttons correspond to which heading. 
 
Similarly, the check box is positioned after the group title heading, between it and the title of the next palette.  The check box has no label, so it is impossible for a screen reader user to gauge its purpose or which palette it belongs to. 
 
Finally, the rendering of the button rows themselves is also likely to cause confusion.  It relies on a separate component, MathButtonsGroup, which exposes a readOnly boolean property.  This property is passed down to the child MathButton component, which is responsible for generating the mark-up for the < button > element.  When readOnly is true (as is the case in this dialog), a disabled attribute is applied to the button. 
 
The upshot of this is that for each palette, users will encounter a list of buttons, all of which are disabled.  However, some of them include their shortcut keys in their accessible labels such as: “stack below (shortcut is : Ctrl+Enter)”.  This will leave users wondering why shortcut keys are indicated for buttons that can’t actually be used. 
 
In other cases, the shortcut isn’t a key combination, but a series of characters which should be typed into the Math editor area.  For example, “Perpendicular (shortcut is : backslash perp⏎)”.  But there is no editor in this dialog.

**Comment by johnhbenetech** [comment]
Suggested Solution 

Check Box Label 
 
All form controls must have an accessible label, and the check boxes in this dialog are no exception.  It’s currently unclear what the check box is supposed to do, but presumably it is used to indicate whether a particular palette button group should be enabled for the new problem set. 
 
As such, each check box needs a unique and informative label such as: “Enable Operators”.  This can be accomplished by giving the < input > element a unique ID, and then following it with a < label > element with a “for” attribute referencing the DOM ID of the check box. 
 
Help Text 
 
It would be helpful if some sort of descriptive text was provided, which indicated how the check box and list of buttons relate to each other.  This could be as simple as: 
 
• Heading level 2: “Geometry” 
• Text: “The Geometry palette includes these buttons:”
• … list of Geometry buttons … 
• Check box: “Enable Geometry” 
 
Button Group Rendering 
 
Modify the MathButton component so that as a minimum, the shortcut text is not included when the readOnly property is set.  Going further, it may make sense to avoid rendering buttons in this dialog at all, given that they are only for informational purposes and can never actually be enabled.  As such, a list of non-interactive textual elements would be better.

**Comment by rupeshparab** [comment]
@johnhbenetech 
I need the helper text for palette categories and also the position to display it in the palette modal

I have implemented rest of the feedback

**Comment by johnhbenetech** [comment]
@alexrcabral @benetechMatt can you look at the above? This is related to the dialogue that pops up when you create a new problem set to pick which buttons you want to be available.

Sina has recommended that each section get some helper text to explain what the checkbox actually does. Can you provide this text and mockup where we would display it?

**Comment by benetechMatt** [comment]
Display the helper text below the category title.  I assume it will be a <p> and therefore visually different than the category title.

Edit
"The edit palette has non-mathematical buttons."

Operators
"The Operators palette has basic mathematical operations buttons."

Notations
"Additional mathematical operators."

Geometry
"Buttons related to to the geometric relation of numbers."

**Comment by jscholes** [comment]
@benetechMatt - the helper text shouldn't repeat the group name which is already present in the heading.  I'd suggest something like:

* Edit: "Non-mathematical buttons."
* Operators: "basic mathematical operations."
* Notations: "Additional mathematical operators."
* Geometry: "related to the geometric relation of numbers."

Given how short these are, they could even just go after a colon in the group name heading:

```
<h2 ...>Edit: Non-mathematical buttons</h2>
```

... and so on.

**Comment by jscholes** [comment]
@rupeshparab - this is looking pretty good.  Some notes:

#### Check Box Labels

The check box label changes based on its state.  A screen reader will announce something like: "Disable Operators check box checked".  This actually makes it sound like the Operators group is disabled, and that I should uncheck the box to enable it.

When I do that though, the label changes to "Enable Operators".  Now I feel like I should check the box to enable the group, and round and round we go.  Ultimately a screen reader user is a bit bewildered as to what is enabled and what isn't.

A screen reader will already indicate the checked state of a check box, so the label text should stay the same.  I think the aim is for the group to be enabled when the box is checked, so the label should be "Enable [groupname]".

#### Button Roles and Shortcuts

The items within each palette still indicate their shortcut keys (or text) in this modal.  They are also still marked up as disabled buttons rather than plain text.  I personally find this confusing as a screen reader user, but would love more input on the visual side from @johnhbenetech, @alexrcabral and @benetechMatt.

#### Missing Button Name

The fourth button in the "Edit" group doesn't 
seem to have a name, just a textual shortcut: "(shortcut is : backslash text{} & Enter to insert)".

**Comment by johnhbenetech** [comment]
@jscholes re: palette. I think the intention was to give the user an idea of:

1. exactly which buttons are included in that group
2. A visual representation of how those buttons will render in the palette if selected

@benetechMatt Do we think we still need to visually render the items as buttons? Would some pure text representation work for the user?

Possible solution is to render the group as an image and use the alt to list the buttons within. Seems a little sloppy, especially since the palette and groups are configurable through a .json definition (and maybe in the future will be a per-user definition) - and so we'd have to come up with some overengineered nonsense to render an image and alt on the fly.

**Comment by jscholes** [comment]
@johnhbenetech - I'm putting the brakes on an image-based approach straight away.  The alt attribute could easily become unmanageably long, and there are tons of issues with using images for significant content which could otherwise be rendered inline.  Not to mention, if this ever does display user-generated stuff, there could be colour contrast issues and all sorts of other nonsense.

I'll wait for more feedback from your side on this one.  I have some ideas for enhancement, such as an aria-describedby on the check boxes to programmatically link them with the palette contents.

**Comment by johnhbenetech** [comment]
@jscholes yep, makes sense. Let us know what you think @benetechmatt

**Comment by benetechMatt** [comment]
Displaying the buttons is important, but they do not have to be HTML buttons. In the future we may allow teachers to even create their own palettes or layouts.
button keyboard shortcuts are not important for this section

**Comment by johnhbenetech** [comment]
@jscholes so if we need to somehow visually render these elements as well as making it digestible for screen readers, what would be your recommendation?

**Comment by jscholes** [comment]
@johnhbenetech - first, every button needs a name.  As I highlighted above, the fourth button in the "Edit" group doesn't seem to have one, just a textual shortcut: "(shortcut is : backslash text{} & Enter to insert)".

After that, the existing list mark-up should remain, but when readOnly is true on MathButtonsGroup:

* the disabled buttons should be replaced with spans (or just text within the `<li>`; and
* the shortcuts should not be rendered at all.

Once the above changes are implemented, we can think about having an accessible description for the check box and any other enhancements.

CC @rupeshparab

**Comment by benetechMatt** [comment]
The mouse pointer switches between a pointer finger, text cursor, and mouse arrow when moving over the math buttons. (using Chrome 77)

Since these are buttons that are not clickable I think the cursor should stay as a mouse pointer.

**Comment by rupeshparab** [comment]
I have fixed the mouse pointer for non clickable math buttons, also the aria-labelledby issue has been recently fixed in #754

**Comment by jscholes** [comment]
@rupeshparab - great.  Only the dynamic check box labels left to clean up. https://github.com/benetech/MathShare/issues/755#issuecomment-546320997

**Comment by rupeshparab** [comment]
@jscholes 
I have removed the dynamic enable/disable from the label

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/de612d02e85dae00d377e7261e7e8e295d260c12.
